remotes/promises.go

```diff
@@ -0,0 +1,213 @@
+package remotes
+
+import (
+	"context"
+	"reflect"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// scheduler is an abstraction over a component capable of running tasks
+// concurrently
+type scheduler interface {
+	// schedule a task and returns a promise notifying completion
+	schedule(func(ctx context.Context) error) promise
+	// ctx returns the context associated with this scheduler
+	ctx() context.Context
+}
+
+// dependency represents an asynchronous operation with its completion channel and its error state
+type dependency interface {
+	Done() <-chan struct{}
+	Err() error
+}
+
+// failedDependency is a dependency already ran to completion with an error
+type failedDependency struct {
+	err error
+}
+
+func (f failedDependency) Done() <-chan struct{} {
+	return doneCh
+}
+
+func (f failedDependency) Err() error {
+	return f.err
+}
+
+// doneDependency is a dependency already ran to completion without error
+type doneDependency struct {
+}
+
+func (doneDependency) Done() <-chan struct{} {
+	return doneCh
+}
+
+func (doneDependency) Err() error {
+	return nil
+}
+
+// doneCh is used internally by doneDependency and failedDependency (it is a closed channel)
+var doneCh = func() chan struct{} {
+	ch := make(chan struct{})
+	close(ch)
+	return ch
+}()
+
+// whenAll wrapps multiple dependencies in a single dependency
+// the result is completed once any dependency completes with an error
+// or once all dependencies ran to completion without error
+func whenAll(dependencies []dependency) dependency {
+	completionSource := &completionSource{
+		done: make(chan struct{}),
+	}
+	go func() {
+		defer close(completionSource.done)
+		cases := make([]reflect.SelectCase, len(dependencies))
+		for ix, dependency := range dependencies {
+			cases[ix] = reflect.SelectCase{
+				Chan: reflect.ValueOf(dependency.Done()),
+				Dir:  reflect.SelectRecv,
+			}
+		}
+		for len(dependencies) > 0 {
+			ix, _, _ := reflect.Select(cases)
+			if err := dependencies[ix].Err(); err != nil {
+				completionSource.err = err
+				return
+			}
+			cases = append(cases[:ix], cases[ix+1:]...)
+			dependencies = append(dependencies[:ix], dependencies[ix+1:]...)
+		}
+	}()
+	return completionSource
+}
+
+// promise is a dependency attached to a scheduler. It allows to schedule continuations
+type promise struct {
+	dependency
+	scheduler scheduler
+}
+
+func (p promise) wait() error {
+	<-p.Done()
+	return p.Err()
+}
+
+// then schedules a continuation task once the current promise is completed.
+// It propagates errors and returns a promise wrapping the continuation
+func (p promise) then(next func(ctx context.Context) error) promise {
+	completionSource := &completionSource{
+		done: make(chan struct{}),
+	}
+	go func() {
+		defer close(completionSource.done)
+		select {
+		case <-p.Done():
+			if err := p.Err(); err != nil {
```

Comment:
nit: `p.Err()` is only used once below, don't think we need to assign it here.

Reply:
Don't really understand what you mean.
This looks at the parent error and if not nil, cancels the child task scheduling an return early (while if no error occurs, we schedule the child task and wait for its completion.